File: fixngo/backend/src/models/orderModel.js
const mongoose = require('mongoose');

const checklistItemSchema = mongoose.Schema(
  {
    key: { type: String, required: true },
    label: { type: String, required: true },
    done: { type: Boolean, default: false },
  },
  { _id: false }
);

const orderSchema = mongoose.Schema(
  {
    user: { type: mongoose.Schema.Types.ObjectId, required: true, ref: 'User' },
    brand: { type: String, required: true },
    model: { type: String, required: true },
    issues: [{ type: String, required: true }],
    total: { type: Number, required: true },
    status: {
      type: String,
      enum: ['pending', 'assigned', 'in_progress', 'completed', 'cancelled'],
      default: 'pending',
    },
    technician: { type: String, default: '' },
    technicianUser: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    dispatchStatus: {
      type: String,
      enum: ['none', 'offered', 'accepted', 'declined'],
      default: 'none',
    },
    customerPhone: { type: String, default: '' },
    serviceAddress: { type: String, default: '' },
    city: { type: String, default: '' },
    pincode: { type: String, default: '' },
    serviceLat: { type: Number, default: null },
    serviceLng: { type: Number, default: null },
    location: {
      type: { type: String, default: 'Point' },
      coordinates: { type: [Number], default: [0, 0] },
    },
    checklist: [checklistItemSchema],
    photos: {
      before: [{ type: String }],
      after: [{ type: String }],
    },
    notes: { type: String, default: '' },
    servicePhotos: [
      {
        url: { type: String, required: true },
        fileName: { type: String, required: true },
        uploadedAt: { type: Date, default: Date.now },
        uploadedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
      },
    ],
    serviceNotes: [
      {
        text: { type: String, required: true },
        addedAt: { type: Date, default: Date.now },
        addedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
      },
    ],
    paymentStatus: { type: String, enum: ['pending', 'collected'], default: 'pending' },
    paymentMethod: { type: String, enum: ['cash', 'card', 'upi'], default: 'cash' },
    stripePaymentIntentId: { type: String, default: '' },
    technicianEarning: { type: Number, default: 0 },
    estimatedDateTime: { type: Date, default: null },
    completedAt: { type: Date, default: null },
    description: { type: String, default: '' },
    statusHistory: [
      {
        status: String,
        note: String,
        at: { type: Date, default: Date.now },
      },
    ],
  },
  { timestamps: true }
);

orderSchema.index({ location: '2dsphere' });

module.exports = mongoose.model('Order', orderSchema);

File: fixngo/apps/customer_app/lib/models/order_model.dart
class OrderModel {
  final String id;
  final String brand;
  final String model;
  final List<String> issues;
  final int total;
  final String status;
  final String? technicianName;
  final String? technicianUser;
  final double? technicianRating;
  final String? technicianPhone;
  final String paymentStatus;
  final String serviceAddress;
  final double? serviceLat;
  final double? serviceLng;
  final List<StatusHistory>? statusHistory;

  OrderModel({
    required this.id,
    required this.brand,
    required this.model,
    required this.issues,
    required this.total,
    required this.status,
    this.technicianName,
    this.technicianUser,
    this.technicianRating,
    this.technicianPhone,
    required this.paymentStatus,
    required this.serviceAddress,
    this.serviceLat,
    this.serviceLng,
    this.statusHistory,
  });

  factory OrderModel.fromJson(Map<String, dynamic> json) {
    return OrderModel(
      id: json['_id'] ?? '',
      brand: json['brand'] ?? '',
      model: json['model'] ?? '',
      issues: List<String>.from(json['issues'] ?? []),
      total: json['total'] ?? 0,
      status: json['status'] ?? 'pending',
      technicianName: json['technicianName'],
      technicianUser: json['technicianUser'] is String 
          ? json['technicianUser'] 
          : (json['technicianUser'] is Map ? json['technicianUser']['_id'] : null),
      technicianRating: (json['technicianRating'] as num?)?.toDouble(),
      technicianPhone: json['technicianPhone'],
      paymentStatus: json['paymentStatus'] ?? 'pending',
      serviceAddress: json['serviceAddress'] ?? '',
      serviceLat: (json['serviceLat'] as num?)?.toDouble(),
      serviceLng: (json['serviceLng'] as num?)?.toDouble(),
      statusHistory: (json['statusHistory'] as List?)
          ?.map((e) => StatusHistory.fromJson(e))
          .toList(),
    );
  }
}

class StatusHistory {
  final String status;
  final String note;
  final DateTime at;

  StatusHistory({required this.status, required this.note, required this.at});

  factory StatusHistory.fromJson(Map<String, dynamic> json) {
    return StatusHistory(
      status: json['status'] ?? '',
      note: json['note'] ?? '',
      at: DateTime.parse(json['at'] ?? DateTime.now().toIso8601String()),
    );
  }
}
